hlpr_feature_extraction/include/objectFeatures.hpp

```diff
@@ -400,5 +469,195 @@ for(size_t j = 0; j < projected_cluster->points.size(); j++)
+	public:
+		HSHistogram(int dim):dim(dim){ }         // empty constructor -> just sets the dimension of the histogram
+		
+	void computeHistogram(PointCloudT &cloud, std::vector<std::vector<uint> > &hist2, std::vector<double> &hist2_double_vector){
+		int cloud_size = cloud.points.size();
+		
+		for(int i = 0; i < cloud_size; i++){
+            int r = (int)(double)(cloud.points[i].r);
+            int g = (int)(double)(cloud.points[i].g);
+            int b = (int)(double)(cloud.points[i].b);
+            
+            // convert the rgb value to HS value
+            std::pair<double, double> HS = rgb2hue(r,g,b);
+            //std::cout<<"Value of hue: "<<std::get<0>(HS)<< "    sat: "<<std::get<1>(HS)<<std::endl;
+            
+			// Get the HS bins
+			double round = (double)360 / dim;              // As the range of Hue is from 0-359
+			int h = (int) (std::get<0>(HS) / round);       // Get a bin for hue from 0-15
+			int s = (int) (std::get<1>(HS) * (dim-1));     // Get a bin for saturation from 0-15
+			
+			//std::cout<<"Value of h: "<<h<< "    s: "<<s<<std::endl;
+            hist2[h][s] = hist2[h][s]+1;
+            
+            //std::cout<<"\tValue after updating: "<<hist2[h][s] << " ";
+            //std::cout<< h << " " << s << std::endl;
+		}
+		
+	   // Convert to a double vector
+        for (int i = 0; i < dim; i++) {
+            for (int j = 0; j < dim; j++) {
+				hist2_double_vector[i * dim + j] = hist2[i][j];
+            }
+        }
+        
+        for (int i = 0; i < hist2_double_vector.size(); i++) {
+            hist2_double_vector[i] /= (double) cloud_size;
+        }
+	}
+		
+	std::pair<double, double> rgb2hue(int r, int g, int b){
+	  const unsigned char max = std::max (r, std::max (g, b));
+	  const unsigned char min = std::min (r, std::min (g, b));
+
+	  float hue;
+	  float sat;
+
+	  const float diff = static_cast <float> (max - min);
+
+	  if (max == 0) // division by zero
+	  {
+		sat = 0;
+		//hue undefined! set to your favorite value (any value with zero saturation will produce black)
+		hue = 0;
+	  }
+	  else
+	  {
+		sat = diff/max;
+
+		if (min == max) // diff == 0 -> division by zero
+		{
+		 sat = 0;
+		 //hue undefined! set to your favorite value (black)
+		 hue = 0;
+		}
+		else
+		{
+		  if      (max == r) hue = 60.f * (      static_cast <float> (g - b) / diff);
+		  else if (max == g) hue = 60.f * (2.f + static_cast <float> (b - r) / diff);
+		  else    hue = 60.f * (4.f + static_cast <float> (r - g) / diff); // max == b
+
+		  if (hue < 0.f) hue += 360.f;
+		}
+	  }
+
+	  //if (sat < ridiculous_global_variables::saturation_threshold && ridiculous_global_variables::ignore_low_sat)
+		//hue = ridiculous_global_variables::saturation_mapped_value; //hackzz oh the hackz
+		
+		// Return both hue and saturation as a pair
+		std::pair <double, double> returnHS = std::make_pair(hue,sat);
+
+	  return returnHS;
+	}
+};
+
+// Helper class to help with colour histograms
+/*class ColorHistogram {
```

Comment:
Is all this commented code useful? If not, let's delete it. If it gets added to the codebase no one will ever remove it.

Reply:
Not sure why this is commented out...

Reply:
Ok, I think HSHistogram replaced ColorHistogram. So, should delete.